rfcs/006-mango-fdb.md

```diff
@@ -0,0 +1,154 @@
+# Mango RFC
+
+- - - -
+name: Formal RFC
+about: Submit a formal Request For Comments for consideration by the team.
+title: ‘Mango JSON indexes in FoundationDB’
+labels: rfc, discussion
+assignees: ‘’
+
+- - - -
+
+[NOTE]: # ( ^^ Provide a general summary of the RFC in the title above. ^^ )
+
+# Introduction
+
+This document describes the data model and indexing management for Mango json indexes in FoundationDB.
+
+## Abstract
+
+This document details the data model for storing Mango indexes.  The basic model is that we would have a namespace for storing defined indexes and then a dedicated namespace per index for the key/values for a given index. Indexes will be updated in the transaction that a document is written to FoundationDB. When an index is created on an existing database, a background task will build the index up to the Sequence that the index was created at.
+
+## Requirements Language
+
+[NOTE]: # ( Do not alter the section below. Follow its instructions. )
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”,
+“SHOULD”, “SHOULD NOT”, “RECOMMENDED”,  “MAY”, and “OPTIONAL” in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+`Sequence`: a 13 byte value formed by combining the current `Incarnation` of the database and the `Versionstamp` of the transaction. Sequences are monotonically increasing even when a database is relocated across FoundationDB clusters. See (RFC002)[LINK TBD]  for a full explanation.
```

Comment:
You will probably be using the tuple encoding for the keys, and the `Incarnation` will therefore be a variable length integer that will be one byte the vast majority of the time (under values of 249 I believe), but will not be strictly speaking 13 bytes.

Reply:
Thanks @ryanworl, yes the incarnation will be a single byte. I just copied and pasted that section from here https://github.com/apache/couchdb-documentation/pull/397/files#diff-5a51b92701c50f4d70a06d3a85daf8e9R40, maybe I should be a little more explicit in what I mean there.

Reply:
I would recommend _not_ making it a fixed-length byte string of length one and instead use an integer. It takes three additional bytes to encode it as a fixed-length byte string if the `Incarnation` is zero and two additional bytes otherwise.

Reply:
Ah thanks @ryanworl, I'll mention that in the other RFC.

Reply:
@ryanworl the motivation behind the use of a fixed-length byte string was to ensure the sortability of Sequences in their external hexadecimal string representation in CouchDB's HTTP API. @davisp reminded me that the variable-length integer encoding in the tuple layer does generate bytes that always sort correctly (of course), so I can see where this is a good improvement. Thanks for pointing it out.